### PR TITLE
[WIP] Improve test coverage for tuple kernels in NN modules

### PR DESCRIPTION
**Pull request by ezyang**
I audited these by going through THNN and looking for kernels with lots
of height/width parameters.  I noticed that 3D kernels tended to have
better coverage.  I did not check if the convolutions were OK (these are
sooo important we probably do have the tests working for them.)

I checked the changes by showing that if you swapped the height/width
parameters, the tests failed.  In some cases, I needed to extend
the test case to catch these problems.  One new field is
expected_output_size, which does exactly what it says on the tin.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
cc @andersenchen @chandlerzuo

File: test/common_nn.py [hunk 1]
```diff
@@ -2972,6 +3001,10 @@ def __call__(self, test_case):
         module = self.constructor(*self.constructor_args)
         input = self._get_input()
 
+        if self.expected_output_size is not None:
+            out = test_case._forward(module, input)
+            test_case.assertTupleEqual(out.size(), self.expected_output_size)
```
Comment (zou3519) [inline test/common_nn.py]: Oh, didn't know this existed... I've been using assertEqual for tuples

**Review by zou3519 — APPROVED** [review]

**Comment by ezyang** [comment]
ASAN failure is real

File: test/common_nn.py (hunk 1 above)
Comment (ezyang) [reply test/common_nn.py]: Yeah the error message for assertEqual is sooo bad. assertTupleEqual is much better

**Pull request closed without merging**